pkg/tfbridge/schema.go

```diff
@@ -1032,83 +1028,115 @@ func propagateDefaultAnnotations(oldInput, newInput resource.PropertyValue, tfs
-			oldStr = tfs.StateFunc()(oldStr)
+			return oldInput, tfs.StateFunc()(oldInput.StringValue()) == newState.StringValue()
 		}
-		return oldStr == newStr
+		return newState, oldInput.StringValue() == newState.StringValue()
 	default:
-		return oldInput.DeepEquals(newInput)
+		return newState, oldInput.DeepEquals(newState)
 	}
 }
 
-func extractInputsFromOutputs(oldInputs, outs resource.PropertyMap,
-	tfs shim.SchemaMap, ps map[string]*SchemaInfo, isRefresh bool) (resource.PropertyMap, error) {
+func extractSchemaInputs(state resource.PropertyValue, tfs shim.Schema, ps *SchemaInfo,
+	rawNames bool) resource.PropertyValue {
 
-	inputs := make(resource.PropertyMap)
-	for name, value := range outs {
-		// If this property is not an input, ignore it.
-		_, sch, _ := getInfoFromPulumiName(name, tfs, ps, false)
-		if sch == nil || (!sch.Optional() && !sch.Required()) {
-			continue
+	contract.Assert(tfs == nil || !tfs.Computed())
+
+	switch {
+	case state.IsArray():
+		etfs, eps := elemSchemas(tfs, ps)
+
+		inputs := make([]resource.PropertyValue, len(state.ArrayValue()))
+		for i, e := range state.ArrayValue() {
+			inputs[i] = extractSchemaInputs(e, etfs, eps, rawNames)
+		}
+		return resource.NewArrayProperty(inputs)
+	case state.IsObject():
+		var tfflds shim.SchemaMap
+		if tfs != nil {
+			if res, isres := tfs.Elem().(shim.Resource); isres {
+				tfflds = res.Schema()
+			}
+		}
+		var psflds map[string]*SchemaInfo
+		if ps != nil {
+			psflds = ps.Fields
 		}
 
-		// Otherwise, copy it to the result.
-		copy, err := copystructure.Copy(value)
-		if err != nil {
-			return nil, err
+		inputs := resource.PropertyMap{}
+		for pulumiName, value := range state.ObjectValue() {
+			_, etfs, eps := getInfoFromPulumiName(pulumiName, tfflds, psflds, rawNames || useRawNames(tfs))
+			if etfs != nil && !etfs.Computed() && (etfs.Required() || etfs.Optional()) {
+				inputs[pulumiName] = extractSchemaInputs(value, etfs, eps, rawNames || useRawNames(tfs))
+			}
 		}
-		inputs[name] = copy.(resource.PropertyValue)
+
+		return resource.NewObjectProperty(inputs)
+	default:
+		return state
 	}
+}
+
+func extractInputsFromOutputs(oldInputs, outs resource.PropertyMap,
+	tfs shim.SchemaMap, ps map[string]*SchemaInfo, isRefresh bool) (resource.PropertyMap, error) {
 
-	// Propagate default annotations from the old inputs.
 	sch := (&schema.Schema{
 		Elem: (&schema.Resource{
 			Schema: tfs,
 		}).Shim(),
 	}).Shim()
 	pss := &SchemaInfo{Fields: ps}
-	propagateDefaultAnnotations(
-		resource.NewObjectProperty(oldInputs), resource.NewObjectProperty(inputs), sch, pss, !isRefresh)
 
-	return inputs, nil
+	var inputs resource.PropertyValue
+	if isRefresh {
+		// If this is a refresh, only extract new values for inputs that are already present.
+		inputs, _ = extractInputs(resource.NewObjectProperty(oldInputs), resource.NewObjectProperty(outs), sch, pss, false)
+	} else {
+		// Otherwise, take a schema-directed approach that fills out all input-only properties.
```

Comment:
> Are there any cases where this could lead to materially "under specifying" the imported resource? What would have to be true for a resource/schema to have a problem with this new approach?

Absolutely. For example, it might be the case that there is a required input that is effectively a discriminant for the shape of the rest of a resource, and particular values require that certain inout properties are set.

Reply:
Sounds like we need to understand the practical impact this will have. Will it be common enough to hit this that we need to allow users to choose whether to do a conservative or liberal import?  Have we tried this on a handful of common resources to understand what it feels like in practice?

Reply:
> Have we tried this on a handful of common resources to understand what it feels like in practice?

Not yet--this is on my list of things to check. Do you have any suggestions for common resources (besides `s3::Bucket` ;))?

Reply:
After offline discussion today, @pgavlin and I agreed that we should not make the changes to the import behaviour here yet.  We need to do a deeper analysis of import independently, as we have significant problems there - but the changes here are quite likely to swing the dial too far in the opposite direction.  We'll need to do a more complete review of how to tackle the issues with `import`.